### ISSUE TITLE
Poetry doesn't use musllinux binary wheel

### ISSUE THREAD

**Issue by sherbang** [python-poetry/poetry#5225]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: python:3.10-alpine docker container
- **Poetry version**: 1.1.13 and 1.2.0a2
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: n/a

## Issue
Poetry isn't finding musllinux binary wheel for the cryptography package that pip finds.  Instead, it compiles from source on Alpine.  Pip finds the binary package and installs cryptography successfully.

[This gist](https://gist.github.com/sherbang/23e346cb15048fbc2f447d139736ced9) contains scripts to reproduce the issue.

**Comment by ashleysommer** [comment]
I've got the same problem, with the orjson package.

orjson now (since v3.6.7) publishes a musllinux binary wheel compatible with Apline Linux.
Installing with `pip3 install "orjson>=3.6.7"` works fine (pip downloads and installs the musllinux bdist wheel), but installing with Poetry instead ignores the bdist and tries to build the package from source.

**Comment by insertjokehere** [comment]
I'm having the same issue with the `grpcio` package - there are musl wheels available that `pip` finds but poetry doesn't - adds about 30 minutes to CI runs

**Comment by abn** [comment]
In theory, the implementation of [PEP 656](https://peps.python.org/pep-0656/) in `packaging` and Poetry's update of the package in #4749 should have fixed the issue. However, this issue can still persist due to `poetry-core`'s vendored `packaging` version being older than what poetry uses. The changes in https://github.com/python-poetry/poetry-core/pull/336 should fix this issue.

You can test this using the following snippet.

```sh
podman run --rm -i --entrypoint sh docker.io/python:3.10-alpine <<EOF
set -xe
apk add gcc libc-dev make libffi-dev curl git
python -m pip install -q git+https://github.com/python-poetry/poetry.git@master
python -m pip install -q git+https://github.com/python-poetry/poetry-core.git@master
poetry new foobar
cd foobar
poetry add cryptography
EOF
```

However, we seem to have a broader issue of Poetry re-using core's vendored packages and thereby being tied to the version there cusing some unexpected issues. While outside the scope of this particular issue, we should handle it in another issue.

**Comment by sherbang** [comment]
Closed as it's working for me in poetry 1.2.0b1 and later
